### PR TITLE
(PDB-2075) Speed unit tests via template copies

### PR DESCRIPTION
**Pull request by rbrw**

### CODE REVIEW

File: ext/travisci/test.sh [hunk 1]
```diff
@@ -5,18 +5,61 @@ set -x
...
+  
+  export PDB_TEST_DB_HOST=127.0.0.1
+  export PDB_TEST_DB_PORT=5432
+  export PDB_TEST_DB_USER=puppetdb
+  export PDB_TEST_DB_ADMIN=pdb_test_admin
+  set +x
+  export PDB_TEST_DB_USER_PASSWORD="$passwd"
+  export PDB_TEST_DB_ADMIN_PASSWORD="$admin_passwd"
+  set -x
+  lein2 test :only puppetlabs.puppetdb.jdbc-test/query-to-vec
```
Comment (ajroetker) [inline ext/travisci/test.sh]: Fyi our jenkins unit tests will also need this change

**Comment by rbrw** [comment]
For some of the changes it may be easier to consult the individual commits, and you'll almost certainly want to ignore whitespace: https://github.com/puppetlabs/puppetdb/pull/1715/files?w=1

**Comment by mullr** [comment]
@rbrw Ok, I've been through the bulk of it, left notes on the commits. One thing I'm missing though is the actual use of the template... where is that? I'd expect to see a `create database with template` or something.

**Comment by mullr** [comment]
I'm :+1: on this. Running locally I see:

```
Ran 382 tests containing 5480 assertions.
0 failures, 0 errors.
```

Does that seem right?

**Comment by wkalt** [comment]
taking a look at this now

File: test-resources/bin/pdb-test-env [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+
```
Comment (wkalt) [inline test-resources/bin/pdb-test-env]: It'd be nice to have a short comment at the top describing what this is for. Same with setup-pdb-pg.

File: test/puppetlabs/puppetdb/config_test.clj [hunk 3]
```diff
@@ -92,8 +93,7 @@
 (deftest garbage-collection
   (let [config-with (fn [base-config]
                       (-> base-config
-                          (update :database merge
-                                  (first tu/available-postgres-configs))
+                          (update :database merge sample-db-config)
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/config_test.clj]: swapping sample-db-config for available-postgres-configs works here, but available-postgres-configs is used in the sync tests so I don't think we'll be able to drop it as we have

**Comment by wkalt** [comment]
This looks pretty good to me generally, and speeds the tests up by a factor of > 2. It breaks the extensions tests with the removal of available-postgres-configs; there might be more issues with those once we get past that one.

Also left a minor note about explanatory comments in the new shell scripts.

File: test/puppetlabs/puppetdb/config_test.clj (hunk 3 above)
Comment (rbrw) [inline test/puppetlabs/puppetdb/config_test.clj]: Right, I'm working on a patch for that too.

File: test-resources/bin/pdb-test-env (hunk 2 above)
Comment (rbrw) [inline test-resources/bin/pdb-test-env]: @wkalt ok, right - for now I just expanded the usage information.  Just shout if that doesn't address your concerns.

Comment (wkalt) [inline test-resources/bin/pdb-test-env]: :+1:

File: ext/travisci/test.sh [hunk 4]
```diff
@@ -1,22 +1,25 @@
...
 ulimit -u 4096
 
+run-unit-tests()
+(
+  pgdir="$(pwd)/test-resources/var/pg"
+  readonly pgdir
+
+  export PGHOST=127.0.0.1
+  export PGPORT=5432
+  test-resources/bin/setup-pdb-pg "$pgdir"
```
Comment (kbarber) [inline ext/travisci/test.sh]: Minor comment: So the test-resources directory is really for clojure related stuff, any reason why these scripts don't just live somewhere in ext/travisci or something like that? At least then all the related items are together also.

**Comment by rbrw** [comment]
I'm working on the related PR for extensions.

**Comment by rbrw** [comment]
I just realized that I need to fix the doc strings for some of the new functions.  i.e. I wrote (defn foo [...] "bar" ...)

**Comment by rbrw** [comment]
There were some rebase conflicts and other changes, so probably warrants another look, but if everything still seems OK (and the tests pass), I think this may be ready now.

**Comment by wkalt** [comment]
@rbrw sorry for getting to this late -- needs a rebase though

**Pull request merged**